### PR TITLE
Hacks site changes

### PR DESCRIPTION
**Pull request by End45**
Some changes to the hacks website:

- Minor changes
  - Actually require required fields
  - Do not mark Main URL as required
  - Remove old hack submission requirements
  - Prevent Discord messages from being used as download links (attachments are ok)
  - Inlucde ongoing machinimas when searching for wip hacks
  - Minor text fixes
- Add an option to silently edit a hack, which does not update its modified date. Useful to fix minor mistakes without bumping the hack to the top of the list.
- Remove the "demo" category. People stated that they didn't understand the difference between that and "wip", and since having a demo is now required to upload hacks, there's no point.
  - The database needs to be updated to change all the "demo" entries to "wip" before this is deployed!

I don't know if there's a way to test this before deploying it. If so, please let me know.

### CODE REVIEW

File: swablu/web.py [hunk 1]
```diff
@@ -403,12 +404,20 @@ async def post(self, hack_id):
...
-        hack['hack_type'] = self.get_body_argument('hack_type', '')
+
+        editing = bool(hack['message_id'])
+
+        try:
+            hack['name'] = self.get_body_argument('name')
+            hack['description'] = self.get_body_argument('description')
+            hack['hack_type'] = self.get_body_argument('hack_type')
+        except MissingArgumentError:
+            return self.redirect(f'/edit/{hack_id}?missing_arg=1')
```
Comment (evandixon) [inline swablu/web.py:415]: Could the name of the missing argument be included in the redirect url?

**Review by evandixon — APPROVED** [review]

File: swablu/web.py (hunk 1 above)
Comment (End45) [reply swablu/web.py:415]: For what purpose? Displaying which argument was missing to the user? What if there's more than one? (Keep in mind that the exception is triggered as soon as one argument is missing, so we don't have the full list if there's multiple).

File: swablu/config.py [hunk 2]
```diff
@@ -190,21 +190,35 @@ def vote_jam(dbcon, jam_key, user_id, hack):
...
-          f"`screenshot2` = %s," \
-          f"`url_main` = %s," \
-          f"`url_discord` = %s," \
-          f"`url_download` = %s," \
-          f"`video` = %s," \
-          f"`hack_type` = %s," \
-          f"`message_id` = %s," \
-          f"`date_updated` = NOW()" \
-          f" WHERE id = %s"
+    if silent:
```
Comment (theCapypara) [inline swablu/config.py:195]: This can be vastly simplified.

File: swablu/roles.py [hunk 3]
```diff
@@ -53,32 +53,24 @@ async def scan_roles():
 def get_hack_type_str(hack_type):
     if hack_type == "balance_hack_wip":
         return "Balance Hack (work in progress)"
-    if hack_type == "balance_hack_demo":
```
Comment (theCapypara) [inline swablu/roles.py:56]: If you remove these options, what happens with hacks that already have this type?

File: swablu/static/filter.js [hunk 4]
```diff
@@ -22,19 +22,16 @@ function _keep(a, b) {
 function refilter() {
     // this is garbage.
```
Comment (theCapypara) [inline swablu/static/filter.js:23]: it really is

**Review by theCapypara — APPROVED** [review]

File: swablu/config.py (hunk 2 above)
Comment (End45) [reply swablu/config.py]: I tried to squeeze an if there when I wrote this but didn't manage to. I just tried again and it seems valid now. Hopefully it's correct.

File: swablu/config.py [hunk 5]
```diff
@@ -192,33 +192,18 @@ def vote_jam(dbcon, jam_key, user_id, hack):
...
+          f"`name` = %s," \
+          f"`description` = %s," \
+          f"`screenshot1` = %s," \
+          f"`screenshot2` = %s," \
+          f"`url_main` = %s," \
+          f"`url_discord` = %s," \
+          f"`url_download` = %s," \
+          f"`video` = %s," \
+          f"`hack_type` = %s," \
+          f"`message_id` = %s" if silent else f"`message_id` = %s,`date_updated` = NOW()" \
```
Comment (theCapypara) [inline swablu/config.py:205]: That works too. I would have done a variable, like this:
```py
date_updated_filter = ""
if not silent:
    date_updated_filter = ",`date_updated` = NOW()"
sql = f"... {date_updated_filter}"
```

Comment (End45) [reply swablu/config.py:205]: Oh, good to know. I'm not very familiar with how these strings work.

**Pull request merged**